update: Rewrite the expressions router Gateway doc (#6437)
Kong/docs.konghq.com: app/_src/gateway/reference/expressions-language/language-references.md

| `\r`            | Carriage return character |
| `\t`            | Horizontal tab character  |
| `\\`            | The `\` character         |
| `\"`            | The `"` character         |

In addition, Expressions supports raw string literals which takes form in `r#"content"#`.
This feature is useful in case you want to write a regex and repeated escape quickly become
tedious to deal with:

For example, if you want to match `http.path` against `/\d+\-\d+` using the regex `~` operator:

With string literals, the predicate will be written as:

```
http.path ~ "/\\d+\\-\\d+"
```

With raw string literals, you can simply write:

```
http.path ~ r#"/\d+\-\d+"#
```

### IpCidr

`IpCidr` represents a range of IP addresses in [Classless Inter-Domain Routing (CIDR)](https://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing) format.
Here are some examples:

IPv4:

```
net.src.ip in 192.168.1.0/24
```

IPv6:
```
net.src.ip in fd00::/8
```

> **Note:** Expressions parser rejects any CIDR literal where the host portion contains
  any non-zero bits. This means that `192.168.0.1/24` won't pass the parser check because
  the intention of the author is unclear.

### IpAddr

> `IpCidr` represents a single IP addresses in [IPv4 Dot-decimal notation](https://en.wikipedia.org/wiki/Dot-decimal_notation),

Reviewer: ```suggestion
`IpAddr` represents a single IP addresses in [IPv4 Dot-decimal notation](https://en.wikipedia.org/wiki/Dot-decimal_notation),
```
typo here?

Reviewer 2: Should be "`IpAddr` represents a single IP address"?

For example, in the following predicate, the constant is parsed as a `Regex`:

```
http.path ~ r#"/foo/bar/.+"#
```

## Operators

Expressions language support rich set of operators that can be performed on various data types.
Here is an overview:

| Operator        | Name                  | Description                                                                           |
|-----------------|-----------------------|---------------------------------------------------------------------------------------|
| `==`            | Equals                | Field value equals to constant value                                                  |
| `!=`            | Not equals            | Field value does not equals to constant value                                         |
| `~`             | Regex match           | Field value matches regex                                                             |
| `^=`            | Prefix match          | Field value starts with constant value                                                |
| `=^`            | Postfix match         | Field value ends with constant value                                                  |

> | `>=`            | Greater than or equal | Field value greater than or equal constant value                                      |

Reviewer: ```suggestion
| `>=`            | Greater than or equal | Field value greater than or equal to constant value                                      |
```

| `>`             | Greater than          | Field value greater than constant value                                               |

> | `<=`            | Less than or equal    | Field value less than or equal constant value                                         |

Reviewer: ```suggestion
| `<=`            | Less than or equal    | Field value less than or equal to constant value                                         |
```

| `<`             | Less than             | Field value less than constant value                                                  |
| `in`            | In                    | Field value is inside constant value                                                  |
| `not in`        | Not in                | Field value is not inside constant value                                              |

> | `contains`      | Contains              | Field value contains constant value                                                   |

Reviewer: Could you add some explanation on the difference between `in`/`not in` and `contains`?

For example, `net.src.ip in 192.168.0.0/24` will only return `true` if the value of `net.src.ip` is within
`192.168.0.0/24`.

### Contains

This operator is used to check the existence of a string inside another string.
For example, `http.path contains "foo"` will return `true` if `foo` can be found anywhere inside `http.path`.
This will match a `http.path` that looks like `/foo`, `/abc/foo`, or `/xfooy`, for example.

### Type and operator semantics

Here are the allowed combination of field types and constant types with each operator:

> **Note:** Rows represents field types where columns represents constant value types.

| LHS/RHS      | `String`                                | `IpCidr`       | `IpAddr` | `Int`                      | `Regex` | `Expression` |
|--------------|-----------------------------------------|----------------|----------|----------------------------|---------|--------------|
| `String`     | `==`, `!=`, `~`, `^=`, `=^`, `contains` | ❌              | ❌        | ❌                          | `~`     | ❌            |
| `IpAddr`     | ❌                                       | `in`, `not in` | `==`     | ❌                          | ❌       | ❌            |

> | `Int`        | ❌                                       | ❌              | ❌        | `==`, `>=`, `>`, `<=`, `<` | ❌       | ❌            |

Reviewer 2: Should we have `!=`?

| `&&`           | And                   | Returns `true` if **both** expressions on the left and right side evaluates to `true` |
| `||`           | Or                    | Returns `true` if **any** expressions on the left and right side evaluates to `true`  |
| `(Expression)` | Parenthesis           | Groups expressions together to be evaluated first                                     |

### Extended descriptions

### In and not in

These operators are used with `IpAddr` and `IpCidr` types to perform an efficient IP list check.
For example, `net.src.ip in 192.168.0.0/24` will only return `true` if the value of `net.src.ip` is within
`192.168.0.0/24`.

### Contains

This operator is used to check the existence of a string inside another string.
For example, `http.path contains "foo"` will return `true` if `foo` can be found anywhere inside `http.path`.
This will match a `http.path` that looks like `/foo`, `/abc/foo`, or `/xfooy`, for example.

### Type and operator semantics

Here are the allowed combination of field types and constant types with each operator:

> > **Note:** Rows represents field types that appears on the left-hand side (LHS) of the predicate where columns represents constant value types that appears on the right-hand side (RHS) of the predicate.

Reviewer 3: ```suggestion
> **Note:** Rows represents field types that display on the left-hand side (LHS) of the predicate where columns represents constant value types that display on the right-hand side (RHS) of the predicate.
```

|----------------------------------|-----------------------------------------|----------------|----------|----------------------------------|---------|--------------|
| `String`                         | `==`, `!=`, `~`, `^=`, `=^`, `contains` | ❌              | ❌        | ❌                                | `~`     | ❌            |
| `IpAddr`                         | ❌                                       | `in`, `not in` | `==`     | ❌                                | ❌       | ❌            |
| `Int`                            | ❌                                       | ❌              | ❌        | `==`, `!=`, `>=`, `>`, `<=`, `<` | ❌       | ❌            |
| `Expression`                     | ❌                                       | ❌              | ❌        | ❌                                | ❌       | `&&`, `      |

{:.note}
> **Notes:** 
  * The `~` operator is described as supporting both `String ~ String` and `String ~ Regex`.
  In reality, `Regex` constant values can only be written as `String` on the right hand side.

>   The presence of `~` operators makes the string value to be treated as a regex.

Reviewer 3: ```suggestion
  The presence of `~` operators treats the string value as a regex.
```

>   Even with the `~` operator, [`String` escape rules described above](#string) still apply and it
>   is almost always easier to use raw string literals for `~` operator as described in the [`Regex` section](#regex).

Reviewer 3: ```suggestion
  Even with the `~` operator, [`String` escape rules described above](#string) still apply and it
  is almost always easier to use raw string literals for the `~` operator as described in the [`Regex` section](#regex).
```
